File: src/application/queries/knowledge_lookup_handler.py
"""Query handler for Knowledge Base lookups.

KnowledgeLookupHandler provides read-side operations: looking up a single
knowledge entry by path, listing categories, and listing topics within a
category.  It delegates file I/O to a KnowledgeReaderProtocol adapter.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Protocol, runtime_checkable

from src.domain.models.knowledge_entry import (
    KnowledgeCategory,
    KnowledgePath,
)

if TYPE_CHECKING:
    from src.domain.models.knowledge_entry import KnowledgeEntry


@runtime_checkable
class KnowledgeReaderProtocol(Protocol):
    """Port for reading knowledge entries from storage."""

    def read_entry(
        self,
        path: KnowledgePath,
        version: str = "current",
    ) -> KnowledgeEntry:
        """Read a single knowledge entry by path.

        Args:
            path: The RLM-addressable knowledge path.
            version: Version to look up (default "current").

        Returns:
            The resolved KnowledgeEntry.

        Raises:
            InvariantViolationError: If the entry is not found.
        """
        ...

    def list_topics(
        self,
        category: KnowledgeCategory,
        tool: str | None = None,
    ) -> tuple[str, ...]:
        """List available topics within a category.

        Args:
            category: The knowledge category.
            tool: Optional tool filter for the TOOLS category.

        Returns:
            Sorted tuple of topic names.
        """
        ...


class KnowledgeLookupHandler:
    """Orchestrates knowledge base lookup operations.

    Parses user-facing path strings into KnowledgePath value objects
    and delegates to a KnowledgeReaderProtocol for actual retrieval.
    """

    def __init__(self, reader: KnowledgeReaderProtocol) -> None:
        self._reader = reader

    def lookup(self, path_str: str, version: str = "current") -> KnowledgeEntry:
        """Look up a knowledge entry by its RLM path string.

        Args:
            path_str: Path string, e.g. "ddd/tactical-patterns".
            version: Version to retrieve (default "current").

        Returns:
            The resolved KnowledgeEntry.

        Raises:
            InvariantViolationError: If the path is invalid or entry not found.
        """
        path = KnowledgePath(raw=path_str)
        return self._reader.read_entry(path, version=version)

    def list_categories(self) -> tuple[str, ...]:
        """Return all available knowledge category values.

        Returns:
            Tuple of category value strings.
        """
        return tuple(c.value for c in KnowledgeCategory)

    def list_topics(
        self,
        category: str,
        tool: str | None = None,
    ) -> tuple[str, ...]:
        """List topics within a category.

        Args:
            category: Category string (must match a KnowledgeCategory value).
            tool: Optional tool filter for the TOOLS category.

        Returns:
            Sorted tuple of topic names from the reader.

        Raises:
            ValueError: If the category string is not a valid KnowledgeCategory.
        """
        try:
            cat = KnowledgeCategory(category)
        except ValueError:
            from src.domain.models.errors import InvariantViolationError

            msg = (
                f"Invalid category '{category}'. "
                f"Valid categories: {[c.value for c in KnowledgeCategory]}"
            )
            raise InvariantViolationError(msg) from None
        return self._reader.list_topics(cat, tool=tool)

File: internal/knowledge/application/knowledge_lookup_handler.go
package application

import (
	"context"
	"fmt"

	"github.com/alto-cli/alto/internal/knowledge/domain"
)

// KnowledgeReader is a handler-local interface for reading knowledge entries.
// Defined where consumed per Go convention.
type KnowledgeReader interface {
	// ReadEntry reads a single knowledge entry by path.
	ReadEntry(ctx context.Context, path domain.KnowledgePath, version string) (domain.KnowledgeEntry, error)

	// ListTopics lists available topics within a category.
	ListTopics(ctx context.Context, category domain.KnowledgeCategory, tool *string) ([]string, error)
}

// KnowledgeLookupHandler orchestrates knowledge base lookup operations.
// It parses user-facing path strings into KnowledgePath value objects
// and delegates to a KnowledgeReader for actual retrieval.
type KnowledgeLookupHandler struct {
	reader KnowledgeReader
}

// NewKnowledgeLookupHandler creates a new KnowledgeLookupHandler.
func NewKnowledgeLookupHandler(reader KnowledgeReader) *KnowledgeLookupHandler {
	return &KnowledgeLookupHandler{reader: reader}
}

// Lookup looks up a knowledge entry by its RLM path string.
func (h *KnowledgeLookupHandler) Lookup(ctx context.Context, pathStr string, version string) (domain.KnowledgeEntry, error) {
	path, err := domain.NewKnowledgePath(pathStr)
	if err != nil {
		return domain.KnowledgeEntry{}, fmt.Errorf("parse knowledge path: %w", err)
	}
	entry, err := h.reader.ReadEntry(ctx, path, version)
	if err != nil {
		return domain.KnowledgeEntry{}, fmt.Errorf("read entry: %w", err)
	}
	return entry, nil
}

// ListCategories returns all available knowledge category values.
func (h *KnowledgeLookupHandler) ListCategories() []string {
	return domain.AllCategoryValues()
}

// ListTopics lists topics within a category.
func (h *KnowledgeLookupHandler) ListTopics(ctx context.Context, category string, tool *string) ([]string, error) {
	cat, err := domain.ParseCategory(category)
	if err != nil {
		return nil, fmt.Errorf("parse category: %w", err)
	}
	topics, err := h.reader.ListTopics(ctx, cat, tool)
	if err != nil {
		return nil, fmt.Errorf("list topics: %w", err)
	}
	return topics, nil
}
